pkg/controller/generic_control.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2019. PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/tidb-operator/pkg/scheme"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// GenericControlInterface is a wrapper to manage typed object that managed by an arbitrary controller
+type TypedControlInterface interface {
+	CreateOrUpdateConfigMap(controller runtime.Object, cm *corev1.ConfigMap) (*corev1.ConfigMap, error)
+	Delete(controller, obj runtime.Object) error
+}
+
+type typedWrapper struct {
+	GenericControlInterface
+}
+
+// NewTypedControl wraps a GenericControlInterface to a TypedControlInterface
+func NewTypedControl(control GenericControlInterface) TypedControlInterface {
+	return &typedWrapper{control}
+}
+
+func (w *typedWrapper) Delete(controller, obj runtime.Object) error {
+	return w.GenericControlInterface.Delete(controller, obj)
+}
+
+func (w *typedWrapper) CreateOrUpdateConfigMap(controller runtime.Object, cm *corev1.ConfigMap) (*corev1.ConfigMap, error) {
+	result, err := w.GenericControlInterface.CreateOrUpdate(controller, cm, MergeConfigMap)
+	if err != nil {
+		return nil, err
+	}
+	return result.(*corev1.ConfigMap), nil
+}
+
+// GenericControlInterface manages generic object that managed by an arbitrary controller
+type GenericControlInterface interface {
+	CreateOrUpdate(controller, obj runtime.Object, mergeFn MergeFn) (runtime.Object, error)
```

Comment:
`Apply` is ambiguous with the [`server-side apply`](https://kubernetes.io/docs/reference/using-api/api-concepts/#server-side-apply) of kubernetes, rename to `CreateOrUpdate`